test/library/draft/Vector/Vector.chpl

```diff
@@ -0,0 +1,1263 @@
+          _extendGeneric(arr);
+          _maybeDecreaseCapacity();
+        }
+        
+        _leave();
+      }
+      return;
+    }
+
+    pragma "no doc"
+    proc _requestCapacity(newCap: int) {
+      if (_capacity >= newCap) then return;
+      if (_capacity == 0) {
+        _capacity = _initialCapacity;
+      }
+      while (_capacity < newCap) {
+        _capacity *= 2;
+      }
+
+      _domain = {0..#_capacity};
+    }
+
+    pragma "no doc"
+    proc _maybeDecreaseCapacity() {
+
+      const threshold = _capacity/2;
+
+      if threshold <= _initialCapacity then
+        return;
+
+      if _size > threshold then
+        return;
+
+      _capacity /= 2; 
+      _domain = {0..#_capacity};
+    }
+    /*
+      Request a change in capacity
+    */
+    proc requestCapacity(newCapacaity: int) {
+      _enter();
+      _requestCapacity(newCapacaity);
+      _leave();
+    }
+
+    pragma "no doc"
+    inline proc ref _extendGeneric(collection) {
+      on this {
+        for item in collection {
+          _append(item);
+        }
+      }
+    }
+
+    /*
+      Extend this vector by appending a copy of each element contained in
+      another list.
+
+      :arg other: A list containing elements of the same type as those
+        contained in this vector.
+      :type other: `list(eltType)`
+    */
+    proc ref extend(other: vector(eltType, ?p)) lifetime this < other {
+      on this {
+        _enter();
+        _requestCapacity(_size + other.size);
+        _extendGeneric(other);
+        _leave();
+      }
+    }
+
+    /*
+      Extend this vector by appending a copy of each element contained in an
+      array.
+
+      :arg other: An array containing elements of the same type as those
+        contained in this vector.
+      :type other: `[?d] eltType`
+    */
+    proc ref extend(other: [?d] eltType) lifetime this < other {
+      on this {
+        _enter();
+        _requestCapacity(_size + d.size);
+        _extendGeneric(other);
+        _leave();
+      }
+    }
+
+    /*
+      Extends this vector by appending a copy of each element yielded by a
+      range.
+
+      .. note::
+
+        Attempting to initialize a vector from an unbounded range will trigger
+        a compiler error.
+
+      :arg other: The range to initialize from.
+      :type other: `range(eltType)`
+    */
+    proc ref extend(other: range(eltType, ?b, ?d)) lifetime this < other {
+      if !isBoundedRange(other) {
+        param e = this.type:string;
+        param f = other.type:string;
+        param msg = "Cannot extend " + e + " with unbounded " + f;
+        compilerError(msg);
+      }
+      on this {
+        _enter();
+        _requestCapacity(_size + other.size);
+        _extendGeneric(other);
+        _leave();
+      }
+    }
+
+    /*
+      Iterate over the elements of this vector.
+
+      :yields: A reference to one of the elements contained in this vector.
+    */
+    iter these() ref {
```

Comment:
Hmm, should these iterators have `on this {`?

Reply:
Will this impact the performance? I'm not sure about the influence though I didn't see one in List.

Reply:
I have been thinking about this and I think we shouldn't have them:

- DefaultRectangularArr's serial iterators don't have them either.
- It feels a bit wrong to me that a `for` loop would yield items from a different locale than where it was being executed on.

P.S. Yes, it would improve the performance if this iterator was invoked on a locale other than the one that owns this Vector.

What do you think?